EIPS/eip-7848.md

```diff
@@ -0,0 +1,122 @@
+---
+eip: 7848
+title: On-chain upgrade signaling
+description: Allows participants to indicate readiness for a client upgrade when producing blocks
+author: William Entriken (@fulldecent)
+discussions-to: https://ethereum-magicians.org/t/eip-xxxx-on-chain-upgrade-signaling/22306
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-12-22
+---
+
+## Abstract
+
+This proposal adds a mechanism for clients to signal their willingness for a network upgrade by including a “reference implementation hash” field in each block. A network upgrade activates only if enough blocks specify that they are using the new software.
+
+## Motivation
+
+Currently, upgrades to Ethereum Mainnet are announced on the ethereum.org blog. This proposal changes that process by activating upgrades based on the consent of network participants.
+
+## Specification
+
+Ethereum consensus clients shall identify with a reference implementation having its equivalent behaviors.
+
+Ethereum consensus clients shall include a `referenceImplementationHash` (specified below) field immediately following the `extraData` field in the block header.
+
+Proposed new Ethereum consensus clients shall initially behave as the then-current network does. If, and only if, a successful upgrade (defined below) is activated will the new behavior take effect.
+
+When a network participant is willing to support a network upgrade, they shall immediately begin using the new software. The blocks they create including the new `referenceImplementationHash` will contribute to the network upgrade activation.
+
+### The Reference Implementation Hash
+
+When proposing a network upgrade, the proposer shall point to a published, feature-complete reference implementation including the new software behaviors.
+
+The **reference implementation hash** is the SHA-256 hash of the tarred source code of this reference implementation.
+
+### Verification
+
+Network participants shall study the reference implementation and decide whether they support an upgrade.
+
+Network participants shall study any specific software they will run and ensure it faithfully implements the behaviors consistent with that referennce implementation.
+
+### Upgrade
+
+Upgrade proposals (often referred to as hard fork EIPs) must specify an upgrade window and threshold. These parameters are implemented in the consensus client:
+
+- `VOTING_WINDOW_BEGIN`: the first block (inclusive) to count votes.
```

Comment:
is this slot based? is this a fixed slot start value that a proposal will add?

Reply:
It is intentionally unspecified in this EIP.

Yes, people using this EIP for upgrades are responsible to choose the start and end blocks however they like.

Reply:
could you please specify it or add <-- TODO --> sections

Reply:
This is intentionally not specified.

If an implementation would like align their `VOTING_WINDOW_BEGIN` and `VOTING_WINDOW_END` choices with a complete epoch, that would be an acceptable implementation.

Probably every implementation would do it that way. But this EIP does not mandate that.